homeassistant/components/powerwall/__init__.py

```diff
@@ -183,44 +185,83 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     return True
 
 
-def _login_and_fetch_base_info(
+async def _login_and_fetch_base_info(
     power_wall: Powerwall, host: str, password: str | None
 ) -> PowerwallBaseInfo:
     """Login to the powerwall and fetch the base info."""
     if password is not None:
-        power_wall.login(password)
-    return call_base_info(power_wall, host)
+        await power_wall.login(password)
+    return await _call_base_info(power_wall, host)
 
 
-def call_base_info(power_wall: Powerwall, host: str) -> PowerwallBaseInfo:
+async def _call_base_info(power_wall: Powerwall, host: str) -> PowerwallBaseInfo:
     """Return PowerwallBaseInfo for the device."""
-    # Make sure the serial numbers always have the same order
-    gateway_din = None
-    with contextlib.suppress(AssertionError, PowerwallError):
-        gateway_din = power_wall.get_gateway_din().upper()
+
+    async def get_gateway_din() -> str:
+        # Make sure the serial numbers always have the same order
+        gateway_din: str
+        with contextlib.suppress(AssertionError, PowerwallError):
+            gateway_din = await power_wall.get_gateway_din()
```

Comment:
If we can't fetch the din, we should fail setup and raise ConfigEntryNotReady

The fallback was only needed for older power wall firmwares we don't support anymore (and no longer work with this integration)